roles/translator/src/downstream_sv1/downstream.rs

```diff
@@ -79,80 +86,155 @@ impl Downstream {
-
+                let is_a = match downstream.safe_lock(|d| !d.authorized_names.is_empty()) {
+                    Ok(is_a) => is_a,
+                    Err(_e) => break,
+                };
                 if is_a && !first_sent && last_notify.is_some() {
                     let message = Self::get_set_difficulty(target.clone());
-                    Downstream::send_message_downstream(downstream_clone.clone(), message).await;
-
-                    let sv1_mining_notify_msg = last_notify.as_ref().unwrap().clone();
-                    let messsage: json_rpc::Message = sv1_mining_notify_msg.try_into().unwrap();
-                    Downstream::send_message_downstream(downstream_clone.clone(), messsage).await;
-                    downstream_clone
-                        .clone()
-                        .safe_lock(|s| {
-                            s.first_job_received = true;
-                        })
-                        .unwrap();
-                    first_sent = true;
+                    Downstream::send_message_downstream(downstream.clone(), message).await;
+
+                    match last_notify.clone() {
+                        Some(sv1_mining_notify_msg) => {
+                            let message: json_rpc::Message = match sv1_mining_notify_msg.try_into()
+                            {
+                                Ok(msg) => msg,
+                                Err(_e) => {
+                                    break;
+                                }
+                            };
+
+                            Downstream::send_message_downstream(downstream.clone(), message).await;
+                            if let Err(_e) = downstream.clone().safe_lock(|s| {
+                                s.first_job_received = true;
+                            }) {
+                                break;
+                            }
+                            first_sent = true;
+                        }
+                        None => break,
+                    }
                 } else if is_a {
-                    let sv1_mining_notify_msg = rx_sv1_notify.clone().recv().await.unwrap();
-                    let messsage: json_rpc::Message = sv1_mining_notify_msg.try_into().unwrap();
-                    Downstream::send_message_downstream(downstream_clone.clone(), messsage).await;
-                    first_sent = true;
+                    select! {
+                        res = rx_sv1_notify.recv().fuse() => {
+                            match res {
+                                Ok(sv1_mining_notify_msg) => {
+                                    let message: json_rpc::Message = match sv1_mining_notify_msg.try_into()
+                                    {
+                                        Ok(msg) => msg,
+                                        Err(_e) => {
+                                            // should this kill the downstream connection or log a status to the main thread and continue
```

Comment:
I think if the tproxy can't send a notify to the downstream it should probably just terminate the connection immediately and let the miner reconnect.